tailfs/remote.go

```diff
@@ -0,0 +1,390 @@
+	As string `json:"who"`
+	// Readers is a list of Tailscale principals that are allowed to read this
+	// share.
+	Readers []string `json:"readers,omitempty"`
+	// Writers is a list of Tailscale principals that are allowed to write to
+	// this share.
+	Writers []string `json:"writers,omitempty"`
+}
+
+// ForRemote is the TailFS filesystem exposed to remote nodes. It provides a
+// unified WebDAV interface to local directories that have been shared.
+type ForRemote interface {
+	// SetFileServerAddr sets the address of the file server to which we
+	// should proxy. This is used on platforms like Windows and MacOS
+	// sandboxed where we can't spawn user-specific sub-processes and instead
+	// rely on the UI application that's already running as an unprivileged
+	// user to access the filesystem for us.
+	SetFileServerAddr(addr string)
+
+	// SetShares sets the complete set of shares exposed by this node. If
+	// AllowShareAs() is true, we will use one subprocess per user to access
+	// the filesystem (see userServer). Otherwise, we will use the file server
+	// configured via SetFileServerAddr.
+	SetShares(shares map[string]*Share)
+
+	// ServeHTTP behaves like the similar method from http.Handler but also
+	// accepts a Permissions map that captures the permissions of the connecting
+	// node.
+	ServeHTTP(permissions Permissions, w http.ResponseWriter, r *http.Request)
+
+	// Close() stops serving the WebDAV content
+	Close() error
+}
+
+func NewFileSystemForRemote(logf logger.Logf) ForRemote {
+	fs := &fileSystemForRemote{
+		logf:        logf,
+		lockSystem:  webdav.NewMemLS(),
+		fileSystems: make(map[string]webdav.FileSystem),
+		userServers: make(map[string]*userServer),
+	}
+	return fs
+}
+
+type fileSystemForRemote struct {
+	logf           logger.Logf
+	lockSystem     webdav.LockSystem
+	fileServerAddr string
+	shares         map[string]*Share
+	fileSystems    map[string]webdav.FileSystem
+	userServers    map[string]*userServer
+	mx             sync.RWMutex
+}
+
+func (s *fileSystemForRemote) SetFileServerAddr(addr string) {
+	s.mx.Lock()
+	s.fileServerAddr = addr
+	s.mx.Unlock()
+}
+
+func (s *fileSystemForRemote) SetShares(shares map[string]*Share) {
+	userServers := make(map[string]*userServer)
+	if AllowShareAs() {
+		// set up per-user server
+		for _, share := range shares {
+			p, found := userServers[share.As]
+			if !found {
+				p = &userServer{
+					logf: s.logf,
+				}
+				userServers[share.As] = p
+			}
+			p.shares = append(p.shares, share)
+		}
+		for _, p := range userServers {
+			go p.runLoop()
+		}
+	}
+
+	fileSystems := make(map[string]webdav.FileSystem, len(shares))
+	for _, share := range shares {
+		fileSystems[share.Name] = webdavfs.New(&webdavfs.Opts{
+			Logf: s.logf,
+			URL:  fmt.Sprintf("http://%v/%v", hex.EncodeToString([]byte(share.Name)), share.Name),
+			Transport: &http.Transport{
+				Dial: func(_, shareAddr string) (net.Conn, error) {
```

Comment:
This is a big function—is it practical to pull some of it out into support methods?